### ISSUE TITLE
Bug: [no-invalid-void-type] How to allow `arg: () => Promise<unknown> | void`?

### ISSUE THREAD

**Issue by awesomeunleashed** [typescript-eslint/typescript-eslint#5679]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.8.3&sourceType=module&code=PTAEGEHsDsBcCcCWA3ApgE1KgHgQwLYAOANqqLJKAM4AWkA7qAK5W4DmZkAZqAAYAK8SPkRVUAHgAqAPlAAfUMkiJ0vUF0jxQAAVgBPQqioBjJIVgBaI8URxg0SBdvJcN9BaUqL+wwChjMFSw1HT0ADKQuOioWgC8oAAUtAwAXKAARpCQpLjQAJSgsbIB0FTZqAB0xJBsSaF5-oHBVHrQxuCu1UzB8QkFRaAA3qAlZaRVNQkARDSoxNVTBQC+jaXBuC1tHfOQ3YWJ-bKCwqKV8EbZaNOzO4u+viCgAKKIsLNa9JoA1lTqtmT0V40PieVSrIIjGzGL4AdSBADVlJh4htWsZEsZOrtYGk+oUjkIRGJxExoF8HPRoLIFKDDkNfKARk1QIRCad9pidt0+gBuBmgRA8BKsk5iAqDfmM5LhSLReAJBBMVANRmM3D0XCvFlssR81UhBgRKIxBJcVxi-krFbGKGwhFIpJo7ZdWANG2IaFwt6IlQJVFbLHdBoPMAAQRBSN4FnOsCY8GgtjY6lJxlgiBgoHQkCM0AA5MFPvAvqBAW8+KTolx-mDRsF3Z6gQBVaCV6v7f3ohKcl24unHIkSUnkhhU+TMFuoKvQDB0iWM2va0VkeLd7G8-mCxIigfiyUGmXG+WK5V79Wa4Lb0561XSo1y03mk+Mq3+W1emjN1vT9COgNc12vh6dpvJ+k7Vn6mztIGAEhqAjZUImfAKEOFLQGo1RiL8bxkAIOqoGoPiJuCdZvk2ZKoe2kEYtBvb4qA-anCS5EjtS47DpSs78gul5iBy0Hroym7Cnhu76resomseKqqmeWo8ag15SqEd4mmaxAWs+vjWqRIHMZSv5Qf+bo6R+enQBBTrQXkQA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1uYDcBDRgBNaPDpWFFS6KAHc+0JpHBgAviBVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Repro Code

```typescript
const click = async (callout: () => Promise<unknown> | void) => {
  const promise = callout();
  if (promise) {
    console.log("this is async")
    await promise;
    console.log("done")
  }
}

// This is fine
click(() => Promise.resolve("hello"))

// This is fine with `callout: () => Promise<unknown> | void`, but gives a TS error with `callout: () => Promise<unknown> | undefined`
click(() => { console.log("hello") })
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-invalid-void-type": "warn"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

I expected to be able to make a function parameter that accepts either a function of type `() => void` or of type `() => Promise<unknown>` in a way that was acceptable to both TS and the lint rules.

### Actual Result

- Since `no-invalid-void-type` disallows `void` in unions in return types, it gives a warning on `arg: () => Promise<unknown> | void`. 
- However, the rule's suggested solution (`arg: () => Promise<unknown> | undefined`) doesn't work here, as TS won't accept a function of type `() => void` in that argument. 
- Switching to `Promise<unknown> | unknown` instead basically just reduces to `() => unknown`, so I lose the `Promise` typing and get an `'await' has no effect on the type of this expression` warning.

### Additional Info

Sorry if the bug template wasn't the right one--I know this is technically intended behavior, but the "new option" template didn't seem to fit what I wanted to ask either.

A few thoughts:

1. If you have a suggestion for "type your function X way instead", I'm certainly open to that.
2. If your suggested solution is "disable the rule for this case", I can accept that, although this doesn't feel like a _super_ obscure use case to me.
3. The `Promise` case is the only one I have run into where it seemed like this mattered, so I feel like my use case could be fixed by allowing `Promise<T> | void` unions. I _think_ it is unique in this case because of the `await` syntax, but I can't say I've fully thought through all of the potential scenarios.

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.38.0` |
| `@typescript-eslint/parser`            | `5.38.0` |
| `TypeScript`                           | `4.8.3` |
| `ESLint`                               | `8.23.0` |
| `node`                                 | `16.16.0` |

**Comment by bradzacher** [comment]
This is intentional because 99.99% of the time when someone writes `Something | void` they didn't mean void and instead meant undefined.

There is this one really rare edge case where you actually did mean `void`. In this instance you should just use a disable comment with a description noting why you're intentionally circumventing the rule.
